bindings/nodejs/hawktracer_client_nodejs.cpp

```diff
@@ -0,0 +1,153 @@
+{
+    _state.stop();
+    Reset();
+}
+
+void Client::set_on_events(const CallbackInfo& info)
+{
+    // maxQueueSize is set to 2 so that even though the first callback is already running there's room for a new callback.
+    // If 2 slots are already filled up, the second callback will pick up the new events with take_events(),
+    // in which case we can ignore napi_queue_full.
+    _state.set_function(ThreadSafeFunction::New(info.Env(),
+                                                info[0].As<Napi::Function>(),
+                                                "HawkTracerClientOnEvent",
+                                                2,
+                                                1));
+}
+
+// This method is called from reader thread, while all other methods are called from js main thread.
+void Client::_notify_new_event()
+{
+    // prevents this from garbage-collected before the callback is finished
+    Ref();
+
+    auto status = _state.use_function(
+        [this](ThreadSafeFunction f)
+        {
+            return f.NonBlockingCall(this, &Client::_convert_and_callback);
+        });
+
+    if (status != napi_ok) {
+        // Callback was not added in the queue, hence no need to increase the reference count.
+        Unref();
+    }
+
+    if (status != napi_ok && status != napi_queue_full) {
+        std::cerr << "Request for callback failed with error code: " << status << std::endl;
+    }
+}
+
+Value Client::_convert_field_value(const class Env& env, const parser::Event::Value& value)
+{
+    switch (value.field->get_type_id()) {
+        case parser::FieldTypeId::UINT8:
+            return Number::New(env, value.value.f_UINT8);
+        case parser::FieldTypeId::INT8:
+            return Number::New(env, value.value.f_INT8);
+        case parser::FieldTypeId::UINT16:
+            return Number::New(env, value.value.f_UINT16);
+        case parser::FieldTypeId::INT16:
+            return Number::New(env, value.value.f_INT16);
+        case parser::FieldTypeId::UINT32:
+            return Number::New(env, value.value.f_UINT32);
+        case parser::FieldTypeId::INT32:
+            return Number::New(env, value.value.f_INT32);
+        case parser::FieldTypeId::UINT64:
+            return Number::New(env, value.value.f_UINT64);
+        case parser::FieldTypeId::INT64:
+            return Number::New(env, value.value.f_INT64);
+        case parser::FieldTypeId::STRING:
+            return String::New(env, value.value.f_STRING);
+        case parser::FieldTypeId::POINTER:
+            return String::New(env, "(pointer)");
+        default:
+            assert(0);
+    }
+}
+
+Object Client::_convert_event(const class Env& env, const LabeledEvent& event)
+{
+    auto o = Object::New(env);
+    for (const auto& it: event.get_values()) {
+        o.Set(it.first, _convert_field_value(env, it.second));
+    }
+    if (!event.get_label_string().empty()) {
+        o.Set("label", String::New(env, event.get_label_string()));
```

Comment:
I'm not sure if that's the right approach; the label might be an actual field in the event, and we probably shouldn't be replacing it. It's done in the native library (client_utils) because client_utils was supposed to be used only for the end-user applications. Since we're building a nodejs bindings (but not end-user application), I'm not sure we should be manipulating fields, as this is something the customer might *not* want.

Reply:
`label` is defined in `core_events.h`, not by the user.
How about doing this only for `HT_CallstackIntEvent` where it's predefined?
When the map files are provided, the user clearly wants to see the string labels, not the int labels.

Reply:
The whole mapping thing is more like a client feature, not a parser feature. I'd recommend to provide bindings only for parser first, and we might extend it to client if needed.